### ISSUE TITLE
Materialized-view shadowable tombstone older than strictly necessary

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#19977]
This issue is about a materialized view where the view key has one column not in the base's key, i.e, the `base_regular_columns_in_view_pk` case

In that case, when we modify that column (let's call it _b_) Scylla needs to delete an old row from the view table, and create a new one. For example, if we had b=0 and set b=1, Scylla needs to delete the row at b=0 and create the row at b=1. This issue is about the **timestamp** of that deletion for the row at b=0. @raphaelsc discovered that in the current implementation, this timestamp is based on the timestamp when b was _previously_ set to its old value (b=0), instead of when b was set to the new value (b=1). He came up with the following example:

```
CREATE TABLE base (a int, b int, c int, PRIMARY KEY (a));
 
CREATE MATERIALIZED VIEW view AS
    SELECT * FROM base
    WHERE a IS NOT NULL AND b IS NOT NULL
    PRIMARY KEY (a, b);
 
INSERT INTO base (a, b, c) VALUES (0, 0, 1) USING TIMESTAMP 0;
# view now has one live row (a,b)=(0,0)
UPDATE base SET b = 1 USING TIMESTAMP 2 WHERE a = 0;
# view now has one live row (a,b)=(0,1)
```
The second SET needs to create a (shadowable) tombstone for the view row (a,b)=(0,0), but inspecting the table with MUTATION_FRAGMENTS (or reading the sstable), we see the tombstone got the timestamp 0 - the timestamp when b was previously set (to 0), instead of the timestamp 2 (when b is now set the 1 causing the deletion.

The claim I want to make in this issue is that **although setting the tombstone timestamp to 0 in this case is not incorrect, we could also set it to 2 and still be correct**.
I'll explain both parts of this sentence below - the current choice is not wrong, but we could have chosen the higher timestamp. 

But before I do that I want to point out that the choice of view-table (shadowable) tombstone is something we took straight out of Cassandra. The relevant Cassandra code in `ViewUpdateGenerator.java` is:

```java
   private long computeTimestampForEntryDeletion(Row existingBaseRow, Row mergedBaseRow)  {
       if (view.hasSamePrimaryKeyColumnsAsBaseTable())  {
          ....
        }
        // has base non-pk column in view pk
        Cell<?> before = existingBaseRow.getCell(view.baseNonPKColumnsInViewPK.get(0));
        assert isLive(before) : "We shouldn't have got there if the base row had no associated entry";
        return deletion.deletes(before) ? deletion.markedForDeleteAt() : before.timestamp();
    }
```

So the timestamp of the old value being changed (`before.timestamp()`) - not the timestamp of the change - is used for the timestamp of the deletion.

The relevant Scylla code is in `view.cc` is (simplified):
```cpp
void view_updates::do_delete_old_entry(const partition_key& base_key, const clustering_or_static_row& existing, const clustering_or_static_row& update, gc_clock::time_point now) {
    ...
        const auto& col_ids = _base_info->base_regular_columns_in_view_pk();
        ...
        } else if (!col_ids.empty()) {
            // We delete the old row using a shadowable row tombstone, making sure that
            // the tombstone deletes everything in the row (or it might still show up).
            // Note: multi-cell columns can't be part of the primary key.
            auto& def = _base->column_at(kind, col_ids[0]);
            auto cell = existing.cells().cell_at(col_ids[0]).as_atomic_cell(def);
            auto ts = cell.timestamp();
```

Again, the timestamp of the old value being changed (cell.timestamp()`) - not the timestamp of the change - is used for the timestamp of the deletion.

Now let's try to prove what the timestamp of this shadowable tombstone **needs** to be:

Consider that the current base row has b=b1 with timestamp **t1**, and we have an update setting b=b2 at timestamp **t2**. We will send a view update to delete the view row b=b1 by creating a _shadowable tombstone_ with timestamp **T**. What does this T need to be?

1. If b=b1 was set at timestamp **t1**, the view row's row marker would be set to the same timestamp t1. (done in view_updates::compute_row_marker()).
2. If the shadowable tombstone is to cause the entire view row to disappear, it just needs to be newer-or-equal than the row marker in the existing view rows (the other cells in the view row don't matter - see `row::compact_and_expire()`). So we must have **`T >= t1`**
3. If at time t2+1 we set b back to its original value b1, this will create a view row marker at t2+1 and we need it to be newer than the shadowable tombstone at T so allow the marker to live and the shadowable tombstone to be ignored so the view row will live. I.e., we need `t2+1 > T'. In other words, ** `T <= t2`.**

Summing up the above conditions we see that _any timestamp_ between `t1` and `t2` (inclusive)  would be a correct choice for the shadowable tombstone's timestamp. Today we choose `t1`, but the proposal in this issue to make it `t2`.

Why is this change interesting or beneficial? For one esoteric reason (that sounds dubious but really affected a user): Today, putting "old" timestamps in memtable confuses things like our tombstone GC algorithms. If an update leaves any of the view's selected columns unchanged it will be written to the view with an "old" timestamp and we have nothing to do against this. But if the user did update all the selected columns, all cells in the view will have new timestamps _except_ the tombstone. The suggestion in this issue will fix that case.

**Comment by nyh** [comment]
Here is cql-pytest code to demonstrate easily, using `MUTATION_FRAGMENTS`, that the timestamp we get today on the shadowable tombstone is `t1`:

```python
# Find a row matching a given row_restriction (e.g., "a=0 AND b=0") in table,
# and return its MUTATION_FRAGMENTS metadata containing things not visible
# in a regular SELECT - such as the tombstones.
# Since MUTATION_FRAGMENTS is a Scylla-only feature, the test is skipped
# when not run on Scylla.
def scylla_inspect_row(cql, table, row_restriction):
    if not is_scylla(cql):
        pytest.skip("Test needs MUTATION_FRAGMENTS, only available on Scylla")
    results = list(cql.execute(f'SELECT * FROM MUTATION_FRAGMENTS({table}) WHERE {row_restriction} ALLOW FILTERING'))
    # currently assume a single row was requested, and that it comes from a
    # single source (cache, memtable or sstable, but not several).
    assert len(results) == 1
    return json.loads(results[0].metadata)

def test_19977(cql, test_keyspace):
    t1=123
    t2=456
    with new_test_table(cql, test_keyspace, 'a int, b int, c int, primary key (a)') as table:
        with new_materialized_view(cql, table, '*', 'a,b', 'a is not null and b is not null') as mv:
            cql.execute(f'INSERT INTO {table} (a,b,c) VALUES (0, 0, 1) USING TIMESTAMP {t1}')
            # view now has row 0,0, c and row marker have timestamp t1
            assert [(0,0,1)] == list(cql.execute(f"SELECT * FROM {mv} WHERE a=0 AND b=0"))
            j = scylla_inspect_row(cql, mv, "a=0 AND b=0")
            assert t1 == j['marker']['timestamp']
            assert t1 == j['columns']['c']['timestamp']
            # Set b=1 at time t2. This will delete the view row b=0 and
            # create a new one with b=1.
            cql.execute(f'UPDATE {table} USING TIMESTAMP {t2} SET b=1 WHERE a=0')
            # view row 0,0 is deleted with tombstone timestamp t1 (why not t2?)
            assert [] == list(cql.execute(f"SELECT * FROM {mv} WHERE a=0 AND b=0"))
            j = scylla_inspect_row(cql, mv, "a=0 AND b=0")
            assert t1 == j['shadowable_tombstone']['timestamp']
            # view row 0,1 is added, its column c has the old timestamp (t1)
            # and row marker has the update's timestamp t2
            assert [(0,1,1)] == list(cql.execute(f"SELECT * FROM {mv} WHERE a=0 AND b=1"))
            j = scylla_inspect_row(cql, mv, "a=0 AND b=1")
            assert t2 == j['marker']['timestamp']
            assert t1 == j['columns']['c']['timestamp']
```

**Comment by nyh** [comment]
I patched view.cc with the following patch

```patch
ndex eea715613b..ed6b602708 100644
--- a/db/view/view.cc
+++ b/db/view/view.cc
@@ -970,21 +971,27 @@ void view_updates::do_delete_old_entry(const partition_key& base_key, const clus
             // the tombstone deletes everything in the row (or it might still show up).
             // Note: multi-cell columns can't be part of the primary key.
             auto& def = _base->column_at(kind, col_ids[0]);
-            auto cell = existing.cells().cell_at(col_ids[0]).as_atomic_cell(def);
-            auto ts = cell.timestamp();
+            auto ts = update.tomb() ?
+                      update.tomb().tomb().timestamp :
+                      update.cells().cell_at(col_ids[0]).as_atomic_cell(def).timestamp();
             if (col_ids.size() > 1) {
                 // This is the Alternator-only support for two regular base
                 // columns that become view key columns. See explanation in
                 // view_updates::compute_row_marker().
                 auto& second_def = _base->column_at(kind, col_ids[1]);
-                auto second_cell = existing.cells().cell_at(col_ids[1]).as_atomic_cell(second_def);
-                auto second_ts = second_cell.timestamp();
+                auto second_ts =
+                    update.tomb() ?
+                    update.tomb().tomb().timestamp :
+                    update.cells().cell_at(col_ids[1]).as_atomic_cell(second_def).timestamp();
                 ts = std::max(ts, second_ts);
                 // Alternator isn't supposed to have more than two col_ids!
                 if (col_ids.size() != 2) [[unlikely]] {
                     utils::on_internal_error(format("Unexpected col_ids length {}", col_ids.size()));
                 }
             }
+            auto cell = existing.cells().cell_at(col_ids[0]).as_atomic_cell(def);
             if (cell.is_live()) {
                 r->apply(shadowable_tombstone(ts, now));
             }
```

This patch nicely "fixes" the shadowable tombstone in the above test to return t2 instead of t1 (`assert t1 == j['shadowable_tombstone']['timestamp']`). With this patch most MV cql-pytest tests I tried passed (I still did not check the Boost testrs, we have many more of those). But unfortunately some of the more elaborate tests in test/alternator failed, conspicuously only the tests dealing with two new key columns (an Alternator-only feature).  Specifically - it seems that the bug #5006 returned (test test_11801_variant2 failed) and also that bug #17119 returned. I still need to investigate what it means and hope it doesn't blow all the air out of the idea that it's fine to use t2 instead of t1 for the shadowable tombstone timestamp and it's just an implementation bug.

**Comment by nyh** [comment]
Note that changing the MV tombstone timestamp became less interesting - and has even less practical importance - after we merged https://github.com/scylladb/scylladb/pull/20446. The timestamp of tombstones in a memtable no longer matters when considering whether to GC some tombstone during compaction, and this was more-or-less the only reason why the potentially-earlier tombstones bothered us.

**Comment by bhalevy** [comment]
See also https://github.com/scylladb/scylla-enterprise/issues/5534#issuecomment-3053449769

**Comment by nyh** [comment]
> See also [scylladb/scylla-enterprise#5534 (comment)](https://github.com/scylladb/scylla-enterprise/issues/5534#issuecomment-3053449769)

What in that comment is relevant here?
This issue is about the choice of timestamp of a **shadowable tombstone**, but the linked comment was about the timestamp of a view **row marker** - or more accurately - whether we can skip writing a row marker at all when we believe an update didn't change it.

**Comment by tgrabiec** [comment]
Implementing this would actually make the problem in https://github.com/scylladb/scylla-enterprise/issues/5534 worse, because producing shadowable tombstones with higher timestamps makes them more likely to overlap with memtables, and thus less likely to be eligible for GC.

**Comment by tarzanek** [comment]
so what is the other option so we can get rid of the workaround we run for them (manual compaction trigger right now every 2 weeks) ?

is there a chance we can at least get this special compaction option be optional on table, so we move away from workaround and have optional way to trigger it in background?

**Comment by tgrabiec** [comment]
Compaction should not be necessary after 9fd312d157604c672cb1172614a0ba95a7108209, so in 2025.4

**Comment by mykaul** [comment]
@tarzanek - I'm closing this, as 2025.4.0 is released.
